### PR TITLE
ensime-vim: refactor and add more features

### PR DESCRIPTION
**Pull request by jvican**
* Start to split the user messages (feedback) and the config from the
  logic

* Use `contextmanagers` to better code abstraction

* Document the most important bits of the logic

* Split logic into internal functions to make their goal clear

* Ensime extends object (pythonic good practice since 2.2)

* Optimize and simplify conditionals

* Add handler for `ArrowTypeInfo` to display types of functions

* Add `EnToggleFullType` to enable to display either full types or
  shallow types

* Use decorator to remove boilerplate code in the neovim plugin

* Add high-level library for threads `threading` instead of the
  deprecated `thread` module

* Make code more readable and mantainable

* Fixed issue when starting client

* Migrating to a more user-friendly structure for the classpath
  configuration in the .config folder.

* Allow both vim and neovim to share the sources of the plugin, hence
  removing boilerplate and inconvenience when editing

* Ensime-shared has been improved considerably and now contains most of
  the logic of the `ensime-vim` plugin

* My credits included in the `ensime-shared` python module

* Improve vim documentation of the ensime-vim plugin (and README)

* Reorder appearance of functions to better readability

* Make `EnDeclaration` go to char correctly

* Add refactoring `rename` in the client side although a lot of
  improvements must be done in the ensime server before getting all the
  goodies of this functionality

* Add up-to-date gif demos

### CODE REVIEW

**Comment by jvican** [comment]
I have decided to PR on `master` because this has been extensively tested. Please, if you find any error let me know or fix it by yourself! :dancer:

**Comment by dubiousdavid** [comment]
Just curious what the status of this PR is?

File: autoload/ensime.vim.py [hunk 1]
```diff
@@ -11,606 +11,10 @@ def EnsimeInitPath():
...
-            findstart = findstart[0]
-        if self.is_scala_file():
-            return self.with_current_client(lambda c: c.complete_func(findstart, base))
-        else:
-            return []
-    def on_receive(self, name, callback):
-        self.with_current_client(lambda c: c.on_receive(name,  callback))
-    def send_request(self, request):
-        self.with_current_client(lambda c: c.send_request(request))
+from ensime_shared.ensime import Ensime
```
Comment (vito-c) [inline autoload/ensime.vim.py]: migration to a shared package probably should have been done in a separate PR so that the features you added are easier to see.

File: ensime_shared/util.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
+import os
+from contextlib import contextmanager
+
+class Util:
```
Comment (vito-c) [inline ensime_shared/util.py]: creating a general utility class usually just leads to too much clutter everything in here is file specific. A better name for this class is FileOperations or FileUtilities

**Comment by vito-c** [comment]
I would suggest breaking this into two PRs to make it easier to see what changes you have added.

File: autoload/ensime.vim.py (hunk 1 above)
Comment (jvican) [inline autoload/ensime.vim.py]: I couldn't submit my PR until support for vim was complete. I did not wanted to repeat all my changes for vim as well. Also, this is meant to be more a general refactoring than several single changes.

File: ensime_shared/util.py (hunk 2 above)
Comment (jvican) [inline ensime_shared/util.py]: I haven't created Util, it was already defined https://github.com/ensime/ensime-vim/blob/master/ensime_launcher/__init__.py#L7

**Comment by jvican** [comment]
Sorry if changes are not clear (they can be found under ensime_shared) but I cannot split this again and support vim as well. As I said before, this was meant to be more of a general refactoring than several single changes. On the way (and because I needed it), I implemented `EnRename` and some more features.

**Comment by yazgoo** [comment]
Thanks @jvican, I tend to agree with @vito-c , splitting this would help.

**Comment by jvican** [comment]
The only remarkable thing that I did is to move the logic to ensime_shared (ensime.py file). I understand that it will help... but I don't have enough time for this in the near future and I would struggle a lot to make all the changes that I did.

**Comment by yazgoo** [comment]
Fair enough. I'll start by testing working with your branch, if anyone can do the same and give feedback, this would be nice.

**Comment by dubiousdavid** [comment]
Here is what I found using Neovim 0.1.1:

Got this error when using omni completion.
```
Error detected while processing function EnCompleteFunc:
line    1:
KeyError('filetype',)
```
Nice to have EnType working with functions, but it was pretty easy to make it hang in such a way that I had to restart Neovim.

**Comment by dubiousdavid** [comment]
Just realized I didn't clear my `.ensime_cache` directory prior to testing, so I'll need to retest after doing that.

**Comment by dubiousdavid** [comment]
Still seeing the same issues.

**Comment by jvican** [comment]
Yes, as I said before, I am fixing the issues with omnicompletion. Let's see what is the problem with EnType.

**Comment by dubiousdavid** [comment]
What's the status of this PR? ensime-vim could really use some better type information.

**Comment by jvican** [comment]
I am trying to close this today.

**Comment by vito-c** [comment]
@jvican for future reference you should make your commits more incremental then they will be easier to break out if needed :)

**Comment by jvican** [comment]
I usually do @vito-c even though I rewrited the git history because some people like only one commit on PRs.

File: README.md [hunk 3]
```diff
@@ -1,96 +1,132 @@
...
 [![Join the chat at https://gitter.im/ensime/ensime-vim](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/ensime/ensime-vim?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
 [![Build Status](https://drone.io/github.com/yazgoo/ensime-vim/status.png)](https://drone.io/github.com/yazgoo/ensime-vim/latest)
 [![Coverage Status](https://coveralls.io/repos/yazgoo/ensime-vim/badge.svg?branch=master&service=github)](https://coveralls.io/github/yazgoo/ensime-vim?branch=master)
 
-ENSIME for the Editor of the Beast (Vim)
-
-# demo
+Vim, the editor of the Beast, has got an **Ensime** plugin to convert your preferred text editor
+into a Java/Scala **IDE**. It's not time for half measures and `ensime-vim` aims at
+becoming your go-to option to easily edit your source code in an efficient and fast way.
```
Comment (yazgoo) [inline README.md]: I think this is a little verbose, couldn't we remove that last sentence ?

Comment (jvican) [inline README.md]: Yep.

File: README.md [hunk 4]
```diff
@@ -1,96 +1,132 @@
...
 
-![alt tag](https://raw.github.com/yazgoo/ensime-vim/master/doc/demo.gif)
+# Demo
 
-# howto
+![First demo](doc/ensime-vim.gif)
+![Second demo](doc/ensime-vim2.gif)
```
Comment (yazgoo) [inline README.md]: This is nice, but I see some issues:

1. having two images uses space
2. EnType result is too fast to see what it does
3. It would be nice to not see ttyrec command
4. we don't see completion example.

I'll try and do a new demo taking this into account.

File: README.md [hunk 5]
```diff
@@ -1,96 +1,132 @@
...
 
-You need `websocket-client` python package:
+# How to get to heaven!
```
Comment (yazgoo) [inline README.md]: I don't think this part is really necessary.

File: README.md (hunk 4 above)
Comment (jvican) [inline README.md]: I am finishing the completion stuff. It will be ready in one hour or so.

File: README.md [hunk 6]
```diff
@@ -1,96 +1,132 @@
...
     $ sudo pip install websocket-client
 
-You should also export your BROWSER variable, for example in your bashrc:
+Then, export your `$BROWSER` variable, for example in your `.bashrc`:
 
-    export BROWSER=firefox
+    $ export BROWSER=firefox
 
-All the following commands should be run from your scala directory.
+*If this variable is pointing to a script, make sure the shebang line (`#!/bin/bash` or whatever)
```
Comment (yazgoo) [inline README.md]: I don't think this is really necessary, what do you think ?

File: README.md (hunk 5 above)
Comment (jvican) [inline README.md]: OK, that was supposed to be a funny comment. If you prefer, hold on until my errors are fixed and submit a commit with all the changes you consider more appropriate.

File: README.md (hunk 6 above)
Comment (jvican) [inline README.md]: Yes, this is really necessary. Otherwise, it won't be executed.

File: README.md (hunk 4 above)
Comment (jvican) [inline README.md]: Ah, and BTW, most of vim plugins uses two images to show the functionality. I wanted to keep that "standard" style.

File: ensime_shared/util.py (hunk 2 above)
Comment (yazgoo) [inline ensime_shared/util.py]: If it gets too big in the future we should add a new class.

File: README.md (hunk 5 above)
Comment (yazgoo) [inline README.md]: OK

File: addon-info.json [hunk 7]
```diff
@@ -1,10 +1,10 @@
 {
   "name" : "ensime-vim",
-  "version" : "0.1",
-  "author" : "Olivier 'yazgoo' Abdesselam",
-  "maintainer" : "Olivier 'yazgoo' Abdesselam",
+  "version" : "0.2",
+  "author" : "Olivier 'yazgoo' Abdesselam, Jorge Vicente Cantero (jvican)",
```
Comment (yazgoo) [inline addon-info.json]: I'ld be enclined to remove all these author fields, or at least point them to the ensime-vim project, and create an AUTHOR.md, would that be fine with you ?

Comment (jvican) [inline addon-info.json]: Sure, no problem!

**Comment by yazgoo** [comment]
Ok, just tell me when you're good to go for mergin.

**Comment by jvican** [comment]
After some hours I didn't find any bug. If you do, please tell me.
  
There is a misbehaviour of the vim omni completion that causes to insert the entire suggested word and doesn't take into account the previous substring. For example, if we request a completion for `Execu` we'll get `ExecuExecutionContext` instead of `ExecutionContext`. I don't know how to solve this and I think it's not related to `ensime-vim` but the backend that uses it.
  
@yazgoo It's your turn to modify the docs. Feel free to remove any funny comment, I am quite prone to them... :laughing:

**Comment by fommil** [comment]
what's up with the build? This doen't look healthy...

```
$ python spec/ensime.py
Traceback (most recent call last):
  File "spec/ensime.py", line 21, in <module>
    import ensime_launcher
ImportError: No module named ensime_launcher
```

**Comment by jvican** [comment]
That was supposed to be set up to execute tests by @yazgoo and I think it's not finished and he got some problems. I already gave my opinion on the future testing of the `ensime-plugin` at https://github.com/ensime/ensime-vim/issues/136.

**Comment by fommil** [comment]
ok cool, so long as we have tests :smile:

thanks for doing all this work! It looks like you're really moving the vim support along.

Can I ask that you put the docs on http://ensime.github.io/editors/vim/ and make the README a short note like https://github.com/ensime/ensime-emacs/blob/master/README.md

We're consolidating all the documentation into one site.

**Comment by jvican** [comment]
Sure! I will do it in the next week as I am moving out in a few days and will be a little bit busy with the `rapture` stuff and my personal issues :smiley_cat:

**Comment by jvican** [comment]
BTW, last commit deletes AUTHORS file as well by request of @fommil, it's better to point out to the github graph [here](https://github.com/ensime/ensime-vim/graphs/contributors).

File: ensime_shared/util.py (hunk 2 above)
Comment (jvican) [inline ensime_shared/util.py]: I frankly think that we should remove the class and only import functions. After all, that's what we want and we want to import only the necessary utils.

**Comment by jvican** [comment]
Any news on this?

**Comment by yazgoo** [comment]
@jvican are you ok for merge ?

**Comment by jvican** [comment]
Yep, everything's fine.

**Comment by jvican** [comment]
Now, you'll have to polish things and get rid of the parts of the docs you don't like. Feel free :+1:

**Comment by yazgoo** [comment]
Awesome,

Thanks for your contribution!
